Sources/AsyncHTTPClient/FileDownloadDelegate.swift

```diff
@@ -0,0 +1,88 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2020 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOHTTP1
+
+/// Handles a streaming download to a given file path, allowing headers and progress to be reported.
+public final class FileDownloadDelegate: HTTPClientResponseDelegate {
+    /// The response type for this delegate: the total count of bytes as reported by the response
+    /// "Content-Length" header (if available) and the count of bytes downloaded.
+    public typealias Response = (totalBytes: Int?, receivedBytes: Int)
+
+    private var totalBytes: Int?
+    private var receivedBytes = 0
+
+    private let handle: NIOFileHandle
+    private let io: NonBlockingFileIO
+    private let reportHeaders: ((HTTPHeaders) -> Void)?
+    private let reportProgress: ((_ totalBytes: Int?, _ receivedBytes: Int) -> Void)?
+
+    private var writeFuture: EventLoopFuture<Void>?
+
+    /// Initializes a new file download delegate.
+    /// - parameters:
+    ///     - path: Path to a file you'd like to write the download to.
+    ///     - pool: A thread pool to use for asynchronous file I/O.
+    ///     - reportHeaders: A closure called when the response headers are available.
+    ///     - reportProgress: A closure called when a body chunk has been downloaded, with
+    ///       the total byte count and download byte count passed to it as arguments.
+    public init(
+        path: String,
+        pool: NIOThreadPool = NIOThreadPool(numberOfThreads: 1),
+        reportHeaders: ((HTTPHeaders) -> Void)? = nil,
+        reportProgress: ((_ totalBytes: Int?, _ receivedBytes: Int) -> Void)? = nil
+    ) throws {
+        self.handle = try NIOFileHandle(path: path, mode: .write, flags: .allowFileCreation())
+        pool.start()
```

Comment:
This doesn't work at the moment, but it should. apple/swift-nio#1572.

Reply:
Would you like me to add a comment that clarifies that the pool shouldn't be started before being passed to the initializer, or should I just wait for that PR to be merged first?

Reply:
I think waiting for that PR to be merged and bumping the minimum NIO version is fine. We're aiming to get a NIO release out this week so we shouldn't be waiting long.

Reply:
I think we shouldn't start the thread pool here. If the user hands you a thread pool, they're responsible for starting & stopping it.

Reply:
That would be great if the `pool` parameter didn't have a default value, the great thing about it though that we don't have to expose pools to our users if they don't want it. Given that apple/swift-nio#1572 was merged, would you mind if the pool is still started here?